### PR TITLE
Namespace is now an option that can be override

### PR DESCRIPTION
**Pull request by naps62**
This allows, for example, for multiple chat's to be used in the same
connection.

### CODE REVIEW

File: lib/errors.js [hunk 1]
```diff
@@ -23,7 +23,8 @@ var errorMap = {
 
   INVALID_LISTENER: ': Listener was not found or invalid',
   INVALID_EVENT: ': Event was not found or invalid',
-  INVALID_MESSAGEEXPIRY: ': messageExpiry can only be a number'
+  INVALID_MESSAGEEXPIRY: ': messageExpiry can only be a number',
+  INVALID_NAMESPACE: 'namespace must be a string'
```
Comment (colinmacdonald) [inline lib/errors.js:27]: Small formatting issue, should be `': namespace'`.

**Comment by colinmacdonald** [comment]
Just a note that you don't have to include the built dist changes, we do this when we tag a new release.

**Comment by colinmacdonald** [comment]
Just needs a test for the error message

Note that the tests will say they failed on travis due to it being a fork, that can be ignored. As long as the tests pass locally we are all set.

File: lib/errors.js (hunk 1 above)
Comment (naps62) [inline lib/errors.js:27]: woops. fixing it

**Comment by naps62** [comment]
@colinmacdonald It seems your latest deploy this week broke this feature
Right now, I have two chat instances on my application, each with a custom namespace. But new messages sent to me are not shown automatically, although they appear after I refresh the page

However, if I change one of the chat instances to use the default namespace ("goinstant/widgets/chat"), then that one will work properly

I couldn't find a reference to that specific string anywhere though

Can this be reverted so that this commit works again? I require this feature for my app, and it was working fine until the latest deploy

Please note, this is independent from the acceptance/rejectane of this pull request. I'm OK with having my own copy of `chat.js`locally if this doesn't get merged, but I still need the chat to work properly when the namespace is different

**Comment by colinmacdonald** [comment]
This couldn't have been working for you before without changes to the regex:
https://github.com/goinstant/chat/blob/master/index.js#L30
https://github.com/goinstant/chat/blob/master/index.js#L269

We check the key name that was set with this regex to make sure it was a chat message key. By passing a custom key namespace it will always fail so remote users won't get the message showing up in the view. This would have to be fixed up to get this to work with the namespace option.

**Comment by naps62** [comment]
Ok great. It's possible I didn't notice this issue at the time, I might have forgotten to test remote messages then, and just send messages with no one else in the room...

Still, I'm now having other unrelated issues, which are preventing the whole goinstant part to work. But that seems to be on my end

Regarding this PR then, how would you feel about changing it to this:
instead of providing a full namespace, the optional parameter would just specify a `[a-zA-Z0-9_]+` string, which would replace the "chat" component on the default key. So we would have "/goinstant/widgets/some_other_thing/", and the regex would still be easy to adapt

Thanks, and sorry for the trouble

**Comment by colinmacdonald** [comment]
We were experiencing higher than normal usage earlier, that may have been what was causing your issues.

I like that idea. I think we would still want the key to ultimately fall on `/messages` for where they are stored, but the namespace could fit in-between the `/chat` and `/messages` key (`/goinstant/widgets/chat/<optional namespace>/messages`). Pretty easy fix.

**Comment by colinmacdonald** [comment]
I'm a bit concerned with how the namespace would affect interacting with the DOM though. i.e. does typing out a message in both inputs, then pressing enter send them both? That would be undesirable. There's definitely tests needed for ensuring that this works properly, which I understand may be outside the scope of this project for you.

**Comment by naps62** [comment]
sure. For now I'll hotfix my local version just for my current app. But once I get the time I'd like to test this more thoroughly

**Comment by colinmacdonald** [comment]
sounds good! :thumbsup:

**Pull request closed without merging**